### ISSUE TITLE
TimeZone parsing issue

### ISSUE THREAD

**Issue by marcoancona** [gautamsi/ews-javascript-api#186]
Thanks for the 0.9.0 release, looks very promising. However I am currently experiencing some issues:
```

[StartTimeZonePropertyDefinition] 		Loading property :		StartTimeZone
[TimeZonePropertyDefinition.LoadPropertyValueFromXmlJsObject] TimeZone info has been updated, Please report any bugs to github
[15:08:06] ServiceLocalException {
  message: 'The time zone definition is invalid or unsupported.',
  stack: '{"message":"The time zone definition is invalid or unsupported.","stack":null,"InnerException":null}\n    at TimeZoneTransitionGroup.Validate (.../node_modules/ews-javascript-api/js/ComplexProperties/TimeZones/TimeZoneTransitionGroup.js:212:27)\n    ...)',
  InnerException: null }
```

Then both `item.Start.MomentDate.toDate()` and `item.End.MomentDate.toDate()` are set to `now()` instead of the correct times from Exchange (whose raw response is reported below)

```
[15:08:06] { __prefix: 's',
  __xmlns: 
   { s: 'http://schemas.xmlsoap.org/soap/envelope/',
     h: 'http://schemas.microsoft.com/exchange/services/2006/types',
     xsi: 'http://www.w3.org/2001/XMLSchema-instance',
     xsd: 'http://www.w3.org/2001/XMLSchema',
     m: 'http://schemas.microsoft.com/exchange/services/2006/messages',
     t: 'http://schemas.microsoft.com/exchange/services/2006/types' },
  Header: 
   { __prefix: 's',
     ServerVersionInfo: 
      { __prefix: 'h',
        __type: 'ServerVersionInfo',
        MajorVersion: '14',
        MinorVersion: '3',
        MajorBuildNumber: '319',
        MinorBuildNumber: '2',
        Version: 'Exchange2010_SP2' } },
  Body: 
   { __prefix: 's',
     GetItemResponse: 
      { __prefix: 'm',
        __type: 'GetItemResponse',
        ResponseMessages: 
         { __prefix: 'm',
           __type: 'ResponseMessages',
           GetItemResponseMessage: 
            { __prefix: 'm',
              __type: 'GetItemResponseMessage',
              ResponseClass: 'Success',
              ResponseCode: 'NoError',
              Items: 
               { __prefix: 'm',
                 __type: 'Items',
                 CalendarItem: 
                  { __prefix: 't',
                    __type: 'CalendarItem',
                    ItemId: 
                     { __prefix: 't',
                       __type: 'ItemId',
                       Id: 'SOME ID',
                       ChangeKey: 'SOME KEY' },
                    Subject: 'SOME TEXT',
                    Sensitivity: 'Normal',
                    Body: { __prefix: 't', __type: 'Body', BodyType: 'Text' },
                    DateTimeReceived: '2017-01-08T12:46:31Z',
                    Importance: 'Normal',
                    ReminderIsSet: 'false',
                    ReminderMinutesBeforeStart: '15',
                    EffectiveRights: 
                     { __prefix: 't',
                       __type: 'EffectiveRights',
                       CreateAssociated: 'false',
                       CreateContents: 'false',
                       CreateHierarchy: 'false',
                       Delete: 'true',
                       Modify: 'true',
                       Read: 'true' },
                    LastModifiedTime: '2017-06-18T21:21:06Z',
                    UID: 'SOME UID',
                    Start: '2017-01-25T16:15:00Z',
                    End: '2017-01-25T18:45:00Z',
                    OriginalStart: '2017-01-25T16:15:00Z',
                    IsAllDayEvent: 'false',
                    Location: 'SOME TEXT',
                    IsRecurring: 'true',
                    CalendarItemType: 'Exception',
                    MyResponseType: 'Organizer',
                    Organizer: 
                     { __prefix: 't',
                       __type: 'Organizer',
                       Mailbox: 
                        { __prefix: 't',
                          __type: 'Mailbox',
                          Name: 'John Smith',
                          EmailAddress: 'mail@company.com',
                          RoutingType: 'SMTP',
                          MailboxType: 'Mailbox' } },
                    TimeZone: '(UTC+01:00) Amsterdam, Berlin, Bern, Rome, Stockholm, Vienna',
                    AppointmentState: '1',
                    StartTimeZone: 
                     { __prefix: 't',
                       __type: 'StartTimeZone',
                       Name: '(UTC+01:00) Amsterdam, Berlin, Bern, Rome, Stockholm, Vienna',
                       Id: 'W. Europe Standard Time',
                       Periods: 
                        { __prefix: 't',
                          __type: 'Periods',
                          Period: 
                           [ { __prefix: 't',
                               __type: 'Period',
                               Bias: '-PT1H',
                               Name: 'Standard',
                               Id: 'trule:Microsoft/Registry/W. Europe Standard Time/1-Standard' },
                             { __prefix: 't',
                               __type: 'Period',
                               Bias: '-PT2H',
                               Name: 'Daylight',
                               Id: 'trule:Microsoft/Registry/W. Europe Standard Time/1-Daylight' } ] },
                       TransitionsGroups: 
                        { __prefix: 't',
                          __type: 'TransitionsGroups',
                          TransitionsGroup: 
                           { __prefix: 't',
                             __type: 'TransitionsGroup',
                             Id: '0',
                             RecurringDayTransition: 
                              [ { __prefix: 't',
                                  __type: 'RecurringDayTransition',
                                  To: 
                                   { __prefix: 't',
                                     __type: 'To',
                                     Kind: 'Period',
                                     To: 'trule:Microsoft/Registry/W. Europe Standard Time/1-Daylight' },
                                  TimeOffset: 'PT2H',
                                  Month: '3',
                                  DayOfWeek: 'Sunday',
                                  Occurrence: '-1' },
                                { __prefix: 't',
                                  __type: 'RecurringDayTransition',
                                  To: 
                                   { __prefix: 't',
                                     __type: 'To',
                                     Kind: 'Period',
                                     To: 'trule:Microsoft/Registry/W. Europe Standard Time/1-Standard' },
                                  TimeOffset: 'PT3H',
                                  Month: '10',
                                  DayOfWeek: 'Sunday',
                                  Occurrence: '-1' } ] } },
                       Transitions: 
                        { __prefix: 't',
                          __type: 'Transitions',
                          Transition: 
                           { __prefix: 't',
                             __type: 'Transition',
                             To: { __prefix: 't', __type: 'To', Kind: 'Group', To: '0' } } } } } } } } } } }
```

**Comment by marcoancona** [comment]
Another, different case:

```
[StartTimeZonePropertyDefinition] 		Loading property :		StartTimeZone
[TimeZonePropertyDefinition.LoadPropertyValueFromXmlJsObject] TimeZone info has been updated, Please report any bugs to github
Error: TimeZoneInfo->FromZoneName : Invalid mapping data
    at Function.TimeZoneInfo.FindSystemTimeZoneById (.../node_modules/ews-javascript-api/js/TimeZoneInfo.js:174:23) at ...)
```

for
```

{ __prefix: 'm',
                __type: 'GetItemResponseMessage',
                ResponseClass: 'Success',
                ResponseCode: 'NoError',
                Items: 
                 { __prefix: 'm',
                   __type: 'Items',
                   CalendarItem: 
                    { __prefix: 't',
                      __type: 'CalendarItem',
                      ItemId: 
                       { __prefix: 't',
                         __type: 'ItemId',
                         Id: 'SOME KEY',
                         ChangeKey: 'SOME KEY' },
                      Subject: 'SOME TEXT',
                      Sensitivity: 'Normal',
                      Body: { __prefix: 't', __type: 'Body', BodyType: 'Text' },
                      DateTimeReceived: '2017-06-18T15:44:43Z',
                      Importance: 'Normal',
                      ReminderIsSet: 'false',
                      ReminderMinutesBeforeStart: '15',
                      EffectiveRights: 
                       { __prefix: 't',
                         __type: 'EffectiveRights',
                         CreateAssociated: 'false',
                         CreateContents: 'false',
                         CreateHierarchy: 'false',
                         Delete: 'true',
                         Modify: 'true',
                         Read: 'true' },
                      LastModifiedTime: '2017-06-18T15:51:01Z',
                      UID: 'SOME UID',
                      Start: '2017-06-23T13:00:00Z',
                      End: '2017-06-23T14:00:00Z',
                      IsAllDayEvent: 'false',
                      IsRecurring: 'false',
                      CalendarItemType: 'Single',
                      MyResponseType: 'Organizer',
                      Organizer: 
                       { __prefix: 't',
                         __type: 'Organizer',
                         Mailbox: 
                          { __prefix: 't',
                            __type: 'Mailbox',
                            Name: 'A Name',
                            EmailAddress: 'mail@company.com',
                            RoutingType: 'SMTP',
                            MailboxType: 'Mailbox' } },
                      TimeZone: 'UTC',
                      AppointmentState: '0',
                      StartTimeZone: 
                       { __prefix: 't',
                         __type: 'StartTimeZone',
                         Name: 'UTC',
                         Id: 'tzone://Microsoft/Utc',
                         Periods: 
                          { __prefix: 't',
                            __type: 'Periods',
                            Period: 
                             { __prefix: 't',
                               __type: 'Period',
                               Bias: 'PT0H',
                               Name: 'UTC rule',
                               Id: 'tzrule://Microsoft/UtcRule' } },
                         TransitionsGroups: 
                          { __prefix: 't',
                            __type: 'TransitionsGroups',
                            TransitionsGroup: 
                             { __prefix: 't',
                               __type: 'TransitionsGroup',
                               Id: '0',
                               Transition: 
                                { __prefix: 't',
                                  __type: 'Transition',
                                  To: 
                                   { __prefix: 't',
                                     __type: 'To',
                                     Kind: 'Period',
                                     To: 'tzrule://Microsoft/UtcRule' } } } },
                         Transitions: 
                          { __prefix: 't',
                            __type: 'Transitions',
                            Transition: 
                             { __prefix: 't',
                               __type: 'Transition',
                               To: { __prefix: 't', __type: 'To', Kind: 'Group', To: '0' } } } } } } },
```

**Comment by gautamsi** [comment]
how did you get `tzone://Microsoft/Utc` value in there. What version you are targetting (EWS version vs actual version of exchange)

for first error, remove all lines from `Validate` function (leave validate function) at .../node_modules/ews-javascript-api/js/ComplexProperties/TimeZones/TimeZoneTransitionGroup.js:212:27

**Comment by marcoancona** [comment]
I am targeting (both EWS version and actual version) Exchange 2010. The server version is 14.3.319.4002.
It looks like others had also problem parsing the timezone (https://stackoverflow.com/a/34883954/3038460)

> I tried tzlocal library approach. It fails for many users because the Id element of the XML was 1. empty OR 2. some random string. ex: "tzone://Microsoft/Utc", "GMT+00:00", "India Time" OR 3. contains Unicode characters that pytz and tzlocal cannot handle. I realised that I can not rely on the ID, it can be any random string and the right thing to do is to parse the timezone definition. I ended up writing a parser for the timezone definition provided in the EWS response.

**Comment by gautamsi** [comment]
I have pushed 0.9.1. it should work given this wired "tzone://..." data. in case you get more like this you can map them by calling `TimeZoneInfo.AddTimeZoneMapping("tzone://value", "their windows/c# counterpart")`

I am planning to add helper library to get updated data from server so that no more this issue. I found that data from Windows Server (timezone and transition rules) can be fetched from Exchange Server, I figured that I can also make such pack of data using PowerShell on my windows box. I may have to completely rewrite the logic of TimeZone conversion.

**Comment by gautamsi** [comment]
this was supposed to be closed in 0.91. looks like I am still using this milestone. changing milestone to 0.9.2
